src/Http/Http/src/Timeouts/RequestTimeoutsMiddleware.cs

```diff
@@ -0,0 +1,94 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Http.Timeouts;
+
+internal sealed class RequestTimeoutsMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly IOptionsMonitor<RequestTimeoutOptions> _options;
+
+    public RequestTimeoutsMiddleware(RequestDelegate next, IOptionsMonitor<RequestTimeoutOptions> options)
+    {
+        _next = next;
+        _options = options;
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+        var timeoutMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutAttribute>();
+        var policyMetadata = endpoint?.Metadata.GetRequiredMetadata<RequestTimeoutPolicy>();
+
+        var options = _options.CurrentValue;
+
+        if (timeoutMetadata is null && policyMetadata?.Timeout is null && options.DefaultTimeout is null)
+        {
+            return _next(context);
+        }
+
+        var disableMetadata = endpoint?.Metadata.GetRequiredMetadata<DisableRequestTimeoutAttribute>();
+        if (disableMetadata is not null)
+        {
+            return _next(context);
+        }
+
+        TimeSpan? timespan;
+        if (policyMetadata is not null)
+        {
+            timespan = policyMetadata.Timeout;
+        }
+        else if (timeoutMetadata is not null)
+        {
+            if (timeoutMetadata.Timeout is not null)
+            {
+                timespan = timeoutMetadata.Timeout.Value;
+            }
+            else
+            {
+                if (options.Policies.TryGetValue(timeoutMetadata.PolicyName!, out var policy))
+                {
+                    timespan = policy.Timeout;
+                }
+                else
+                {
+                    // policy does not exists. what to do now!?
+                    throw new Exception();
+                }
+            }
+        }
+        else
+        {
+            timespan = options.DefaultTimeout;
+        }
+
+        if (timespan is null || timespan == Timeout.InfiniteTimeSpan)
+        {
+            return _next(context);
+        }
+
+        return SetTimeoutAsync();
+
+        async Task SetTimeoutAsync()
+        {
+            var originalToken = context.RequestAborted;
+            var timeoutCts = new CancellationTokenSource(timespan.Value);
+            var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(context.RequestAborted, timeoutCts.Token);
```

Comment:
I think we need to get this down to 0 (amortized). We need to use an `ObjectPool<CancellationTokenSource>` or (https://github.com/dotnet/aspnetcore/blob/4535ea1263e9a24ca8d37b7266797fe1563b8b12/src/Shared/CancellationTokenSourcePool.cs). This combined with RegisterForDispose should get the allocations per request down (removes the async state machine).

Reply:
We could use [IPersistentState](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.connections.features.ipersistentstatefeature?view=aspnetcore-7.0) to store the CTS on the `HttpContext`. It will be cached along with the context.

That avoids any locking and contention from a shared pool.

Reply:
Nice. We can try that.

Reply:
I'm a little confused on what to do here exactly! Could you elaborate more please?

Reply:
@Tratcher Any guidance on this thread? I would like to do this one in this PR if I can.

Reply:
get the benchmarks working first, then we can measure the impact of this change

Reply:
Sure. Could you please point me to a benchmark sample and what do I need to benchmark exactly? I'm thinking a pipeline with and without the request timeouts middleware?